pcsx2/GS/Renderers/OpenGL/GSDeviceOGL.h

```diff
@@ -181,17 +183,33 @@ class GSDeviceOGL final : public GSDevice
 	using OMDepthStencilSelector = GSHWDrawConfig::DepthStencilSelector;
 	using OMColorMaskSelector = GSHWDrawConfig::ColorMaskSelector;
 
-	struct alignas(32) MiscConstantBuffer
+	struct ProgramSelector
 	{
-		GSVector4i EMOD_AC;
+		VSSelector vs;
+		GSSelector gs;
+		PSSelector ps;
 
-		MiscConstantBuffer() { memset(this, 0, sizeof(*this)); }
+		__fi bool operator==(const ProgramSelector& p) const { return vs.key == p.vs.key && gs.key == p.gs.key && ps.key == p.ps.key; }
+		__fi bool operator!=(const ProgramSelector& p) const { return vs.key != p.vs.key || gs.key != p.gs.key || ps.key != p.ps.key; }
+	};
+
+	struct ProgramSelectorHash
+	{
+		__fi std::size_t operator()(const ProgramSelector& p) const noexcept
+		{
+			std::size_t h = 0;
+			HashCombine(h, p.vs.key, p.gs.key, p.ps.key);
+			return h;
+		}
 	};
 
 	static int m_shader_inst;
 	static int m_shader_reg;
 
 private:
+	// Increment this constant whenever shaders change, to invalidate user's program binary cache.
+	static constexpr u32 SHADER_VERSION = 1;
```

Comment:
It looks like you are hashing the actual shader content but just to make sure, this is just so that old garbage isn't taking up space in the shader cache right?  If I forgot to update this when editing a shader, the worst that could happen is my shader cache gets bigger than usual?

Reply:
Correct.